src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1390 @@
+          signature
+
+ Outputs: a list of expressions
+
+ Purpose: for each expression that is a string or that is at a position with
+          an 'S' character in the signature, we declare a new `cprover_string`
+          whose contents is deduced from the expression and replace the
+          expression by this cprover_string in the output list;
+          in the other case the expression is kept as is for the output list.
+
+\*******************************************************************/
+
+exprt::operandst string_refine_preprocesst::process_arguments(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const exprt::operandst &arguments,
+  const source_locationt &location,
+  const std::string &signature)
+{
+  exprt::operandst new_arguments;
+
+  for(std::size_t i=0; i<arguments.size(); i++)
+  {
+    exprt arg=arguments[i];
+    auto it=string_builders.find(arg);
+    if(it!=string_builders.end())
+      new_arguments.push_back(it->second);
+    else
+    {
+      if(i<signature.length() && signature[i]=='S')
+      {
+        while(arg.id()==ID_typecast)
+          arg=arg.op0();
+        if(!is_java_string_type(arg.type()))
+          arg=typecast_exprt(arg, jls_ptr);
+      }
+      exprt arg2=make_cprover_string_assign(goto_program, i_it, arg, location);
+      new_arguments.push_back(arg2);
+    }
+  }
+  return new_arguments;
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::function_signature
+
+  Inputs: the name of a cprover string function
+
+ Outputs: a signature string
+
+ Purpose: if the signature of the given function is defined, return it
+          otherwise return an empty string
+
+\*******************************************************************/
+
+std::string string_refine_preprocesst::function_signature(
+  const irep_idt &function_id)
+{
+  auto it=signatures.find(function_id);
+  if(it!=signatures.end())
+    return it->second;
+  else return "";
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::replace_string_calls
+
+  Inputs: a function in a goto_program
+
+ Purpose: goes through the instructions, replace function calls to string
+          function by equivalent instructions using functions defined
+          for the string solver, replace string literals by string
+          expressions of the type used by the string solver
+
+\*******************************************************************/
+
+void string_refine_preprocesst::replace_string_calls(
+  goto_functionst::function_mapt::iterator f_it)
+{
+  goto_programt &goto_program=f_it->second.body;
+
+  Forall_goto_program_instructions(i_it, goto_program)
+  {
+    if(i_it->is_function_call())
+    {
+      code_function_callt &function_call=to_code_function_call(i_it->code);
+      for(auto arg : function_call.arguments())
```

Comment:
Oh, I'm sorry, but not knowing the precise rules for `auto` just yet: are you sure you end up with a reference and not with a copy? In the latter case you can obviously modify it, but it won't have much of an effect... I'd suggest `auto &arg` to be on the safe side.